### PR TITLE
Use uevent drive sync only in node startup on dynamic drive discovery

### PR DESCRIPTION
**Pull request by balamurugana**
Signed-off-by: Bala.FA <bala@minio.io>

### CODE REVIEW

File: pkg/node/uevent.go [hunk 1]
```diff
@@ -170,7 +162,15 @@ func (handler *ueventHandler) syncDrives(ctx context.Context) {
 			return
 		}
 
-		if !handler.updateDrive(ctx, result.Drive, devices) {
+		if handler.updateDrive(ctx, &result.Drive, devices) {
+			if result.Drive.Status.DriveStatus == directcsi.DriveStatusInUse {
```
Comment (Praveenrajmani) [inline pkg/node/uevent.go:166]: can we also look for Ready drives and mount them as well?

File: pkg/node/node.go [hunk 2]
```diff
@@ -93,7 +93,26 @@ func NewNodeServer(ctx context.Context, identity, nodeID, rack, zone, region str
...
+			string(utils.TopologyDriverRack):     rack,
+			string(utils.TopologyDriverZone):     zone,
+			string(utils.TopologyDriverRegion):   region,
+			string(utils.TopologyDriverNode):     nodeID,
+		},
+		dynamicDriveDiscovery: dynamicDriveDiscovery,
+	}
+	if dynamicDriveDiscovery {
+		klog.V(3).Info("Doing initial drive sync up")
+		handler.syncDrives(ctx)
```
Comment (Praveenrajmani) [inline pkg/node/node.go:110]: shouldn't the initial sync be happening even when `dynamicDriveDiscovery` is not enabled? We can  bring this out of the loop.

File: pkg/node/uevent.go [hunk 3]
```diff
@@ -256,12 +269,21 @@ func (handler *ueventHandler) processEvent(ctx context.Context, device *sys.Devi
...
 		if action == uevent.Remove {
 			handler.removeDrive(ctx, drive, devices)
 		} else {
-			handler.updateDrive(ctx, drive, devices)
+			if handler.updateDrive(ctx, drive, devices) {
+				switch result.Drive.Status.DriveStatus {
+				case directcsi.DriveStatusReady, directcsi.DriveStatusInUse:
+					target := filepath.Join(sys.MountRoot, result.Drive.Status.FilesystemUUID)
+					flags := result.Drive.Spec.RequestedFormat.MountOptions
+					if err = sys.MountXFSDevice(result.Drive.Status.Path, target, flags); err != nil {
```
Comment (Praveenrajmani) [inline pkg/node/uevent.go:282]: Can we tag the error to drive conditions, like we were doing earlier? 

```
// Verify mounts
message := ""
if err := d.verifyDriveMount(existingDrive); err != nil {
	message = err.Error()
	klog.V(3).Infof("mounting failed with: %v", err)
}
utils.UpdateCondition(existingDrive.Status.Conditions,
	string(directcsi.DirectCSIDriveConditionInitialized),
	utils.BoolToCondition(message == ""),
	string(directcsi.DirectCSIDriveReasonInitialized),
	message)

updateOpts := metav1.UpdateOptions{
	TypeMeta: utils.DirectCSIDriveTypeMeta(),
}
_, err = driveClient.Update(ctx, existingDrive, updateOpts)
return err
```

**Comment by Praveenrajmani** [comment]
```diff
diff --git a/pkg/client/interface.go b/pkg/client/interface.go
index c006e9f..1408a5c 100644
--- a/pkg/client/interface.go
+++ b/pkg/client/interface.go
@@ -29,6 +29,7 @@ import (
        "k8s.io/apimachinery/pkg/api/meta"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+       "k8s.io/apimachinery/pkg/labels"
        runtime "k8s.io/apimachinery/pkg/runtime"
        "k8s.io/apimachinery/pkg/runtime/schema"
        "k8s.io/apimachinery/pkg/types"
@@ -154,6 +155,18 @@ func (d *directCSIInterface) Get(ctx context.Context, name string, options metav
 
 // List takes label and field selectors, and returns the list of resource objects that match those selectors.
 func (d *directCSIInterface) List(ctx context.Context, opts metav1.ListOptions) (map[string]interface{}, []map[string]interface{}, error) {
+
+       var labelSelector labels.Selector
+       var err error
+       if d.groupVersion.Version == "v1alpha1" || d.groupVersion.Version == "v1beta1" {
+               // old objects DO NOT support labels and label selectors
+               labelSelector, err = labels.Parse(opts.LabelSelector)
+               if err != nil {
+                       return nil, nil, err
+               }
+               opts.LabelSelector = ""
+       }
+
        result, err := d.resourceInterface.List(ctx, opts)
        if err != nil {
                return nil, nil, err
@@ -167,6 +180,11 @@ func (d *directCSIInterface) List(ctx context.Context, opts metav1.ListOptions)
 
        items := []map[string]interface{}{}
        for i := range migratedResult.Items {
+               if !labelSelector.Empty() {
+                       if !labelSelector.Matches(labels.Set(migratedResult.Items[i].GetLabels())) {
+                               continue
+                       }
+               }
                items = append(items, migratedResult.Items[i].Object)
        }
        return migratedResult.Object, items, nil
diff --git a/pkg/converter/drive_upgrade.go b/pkg/converter/drive_upgrade.go
index dced4b6..2b3344d 100644
--- a/pkg/converter/drive_upgrade.go
+++ b/pkg/converter/drive_upgrade.go
@@ -167,9 +167,10 @@ func driveUpgradeV1Beta2ToV1Beta3(unstructured *unstructured.Unstructured) error
        }
 
        v1beta3DirectCSIDrive.TypeMeta = v1beta2DirectCSIDrive.TypeMeta
+       v1beta2DirectCSIDrive.Status.Path = convertV1Beta1DeviceName(v1beta2DirectCSIDrive.Status.Path)
        utils.UpdateLabels(&v1beta3DirectCSIDrive, map[utils.LabelKey]utils.LabelValue{
                utils.NodeLabelKey:       utils.NewLabelValue(v1beta2DirectCSIDrive.Status.NodeName),
-               utils.PathLabelKey:       utils.NewLabelValue(v1beta2DirectCSIDrive.Status.Path),
+               utils.PathLabelKey:       utils.NewLabelValue(filepath.Base(v1beta2DirectCSIDrive.Status.Path)),
                utils.CreatedByLabelKey:  utils.DirectCSIDriverName,
                utils.AccessTierLabelKey: utils.NewLabelValue(string(v1beta2DirectCSIDrive.Status.AccessTier)),
        })
```

**Comment by wlan0** [comment]
id dynamic-discovery is false, provide a suggestion to the users to enable it and post the deprecation notice below that

```
Enable dynamic drive change management using '--dynamic-discovery' flag
This flag will be made default in the next major release version
```

**Review by wlan0 — APPROVED** [review]

**Pull request merged**